solana-programs/staking-bridge/programs/staking-bridge/src/wormhole_utils.rs

```diff
@@ -0,0 +1,292 @@
+use anchor_lang::prelude::*;
+use anchor_lang::{
+    solana_program,
+    solana_program::{
+    program::invoke_signed,
+    instruction::{
+        AccountMeta,
+        Instruction,
+    },
+}};
+use anchor_spl::token::spl_token;
+
+use crate::constant::{
+    ETH_CHAIN_ID,
+    ETH_RECIPIENT_ADDRESS_PADDED_32_BYTES,
+    ETH_AUDIO_TOKEN_ADDRESS_PADDED_32_BYTES,
+    WORMHOLE_CORE_BRIDGE_ID,
+    WORMHOLE_TOKEN_BRIDGE_ID
+};
+use crate::error::StakingBridgeErrorCode;
+use crate::{
+    PostWormholeMessage,
+    PostWormholeMessageData
+};
+
+/**
+ * 1. Verify that we are calling the Wormhole Token Bridge program.
+ * 2. Verify that we are using the Wormhole Core Bridge program.
+ */
+pub fn check_wormhole_programs(
+    accounts: &mut PostWormholeMessage,
+) -> Result<()> {
+    let program_id = &accounts.program_id;
+    let bridge_id = &accounts.bridge_id;
+
+    // 1. Verify that we are calling the Wormhole Token Bridge program.
+    if program_id.key().to_string() != WORMHOLE_TOKEN_BRIDGE_ID.to_string() {
+        return Err(StakingBridgeErrorCode::NotCallingWormholeTokenBridgeProgram.into());
+    }
+    // 2. Verify that we are using the Wormhole Core Bridge program.
+    if bridge_id.key().to_string() != WORMHOLE_CORE_BRIDGE_ID.to_string() {
+        return Err(StakingBridgeErrorCode::InvalidWormholeCoreBridgeProgram.into());
+    }
+    Ok(())
+}
+
+/**
+ * Verify PDA ownership of the 'from' token account.
+ * Note that anchor checks for the program ownership of the 'from_owner' account,
+ * i.e. that the owner of the 'from' token account is owned by the program.
+ * This is because we use the account macro with seeds and bump for 'from_owner'.
+ */
+pub fn check_wormhole_token_accounts(
+    accounts: &mut PostWormholeMessage,
+) -> Result<()> {
+    let from = &accounts.from;
+    let from_owner = &accounts.from_owner;
+    let from_account_data = from.data.borrow();
+    let from_account_owner= <anchor_spl::token::spl_token::state::Account as anchor_spl::token::spl_token::state::GenericTokenAccount>
+        ::unpack_account_owner(&from_account_data)
+        .unwrap();
+    if from_account_owner != from_owner.key {
+        return Err(StakingBridgeErrorCode::WormholeTokenAccountNotOwnedByPDA.into());
+    }
+    Ok(())
+}
+
+fn hex_string_to_array(hex_string: &str) -> Vec<u8> {
```

Comment:
can we put this in a utils.rs file and put the wormhole stuff in a file called wormhole.rs? and then call them from lib.rs like you're doing?

Reply:
but also, let's just use this crate instead? https://docs.rs/hex/latest/hex/fn.decode.html